src/main/scala/org/broadinstitute/hail/expr/FunctionRegistry.scala

```diff
@@ -0,0 +1,263 @@
+package org.broadinstitute.hail.expr
+
+import org.broadinstitute.hail.stats._
+import org.broadinstitute.hail.utils._
+import org.broadinstitute.hail.variant.{AltAllele, Genotype, Locus, Variant}
+import org.broadinstitute.hail.expr.HailRep._
+
+import scala.collection.mutable
+
+object FunctionRegistry {
+
+  private val registry = new mutable.HashMap[(String, TypeTag), Fun]
+
+  private def lookupExact(name: String, typ: TypeTag): Option[Fun] =
+    registry.get((name, typ))
+
+  private def lookup(name: String, typ: TypeTag): Option[Fun] =
+    registry.get((name, typ))
+
+  private def bind(name: String, typ: TypeTag, f: Fun) = {
+    lookupExact(name, typ) match {
+      case Some(existingBinding) =>
+        throw new RuntimeException(s"The name, ${ name }, with type, ${ typ }, is already bound as ${ existingBinding }")
+      case None =>
+        registry.put((name, typ), f)
+    }
+  }
+
+  def lookupUnaryMethod(ec: EvalContext)(typ: BaseType, name: String)(xAst: AST): Option[() => Any] =
+    lookup(name, UnaryMethodType(typ)).map {
+      case f: UnaryFun[t, u] => AST.evalCompose(ec, xAst)(f)
+      case f: OptionUnaryFun[t, u] => AST.evalFlatCompose(ec, xAst)(f)
+      case f =>
+        throw new RuntimeException(s"Internal hail error, bad binding in function registry for `$name' with argument type $typ: $f")
+    }
+
+  def lookupUnaryMethodReturnType(typ: BaseType, name: String): Option[Type] =
+    lookup(name, UnaryMethodType(typ)).map(f => f.retType)
+
+  def lookupFun(ec: EvalContext)(name: String, typs: Seq[BaseType])(args: Seq[AST]): Option[() => Any] = {
+    require(typs.length == args.length)
+
+    val funTyp = TypeTag.funTypeFromArgs(typs)
+
+    lookup(name, funTyp).map { fn => (fn, typs) match {
+      case (f: UnaryFun[t, u], Seq(x)) =>
+        AST.evalCompose(ec, args(0))(f)
+      case (f: OptionUnaryFun[t, u], Seq(x)) =>
+        AST.evalFlatCompose(ec, args(0))(f)
+      case (f: BinaryFun[t, u, v], Seq(x, y)) =>
+        AST.evalCompose(ec, args(0), args(1))(f)
+      case (f: Arity3Fun[t, u, v, w], Seq(x, y, z)) =>
+        AST.evalCompose(ec, args(0), args(1), args(2))(f)
+      case (f: Arity4Fun[t, u, v, w, x], Seq(x, y, z, a)) =>
```

Comment:
Yeah, I can change them to `_`. Apparently you can use identifiers too and Scala treats them like `_`.